src/Symfony/Component/HttpClient/ResponseSetMonitor.php

```diff
@@ -0,0 +1,155 @@
+    {
+        $this->client = $client;
+        $this->onHeaders = $onHeaders;
+        $this->onBody = $onBody;
+        $this->onError = $onError;
+        $this->responses = new \SplObjectStorage();
+    }
+
+    /**
+     * Adds a response to the monitored set.
+     *
+     * The response must be created with the same client that was passed to the constructor.
+     */
+    public function add(ResponseInterface $response, callable $onHeaders = null, callable $onBody = null, callable $onError = null): void
+    {
+        $this->responses[$response] = [$onHeaders, $onBody, $onError];
+        $this->tick();
+    }
+
+    /**
+     * Monitors pending responses, moving them forward as network activity happens.
+     *
+     * @param float $timeout The maximum duration of the tick
+     *
+     * @return int The number of responses remaining in the set after the tick
+     */
+    public function tick(float $timeout = 0.0): int
+    {
+        return $this->wait($timeout, false);
+    }
+
+    /**
+     * Completes all pending responses.
+     *
+     * @param float|null $idleTimeout The maximum inactivy timeout before erroring idle responses
+     */
+    public function complete(float $idleTimeout = null): void
+    {
+        $this->wait($idleTimeout, true);
+    }
+
+    /**
+     * Cancels all pending responses.
+     */
+    public function cancel(): void
+    {
+        foreach ($this->responses as $response) {
+            $response->cancel();
+        }
+
+        $this->responses = new \SplObjectStorage();
+    }
+
+    /**
+     * Returns the number of pending responses.
+     */
+    public function count(): int
+    {
+        return \count($this->responses);
+    }
+
+    public function __destruct()
+    {
+        $this->wait(null, true);
+    }
+
+    private function wait(?float $timeout, bool $errorOnTimeout): int
+    {
+        $error = null;
+        $remainingTimeout = $timeout;
+
+        if (!$errorOnTimeout && $remainingTimeout) {
+            $startTime = microtime(true);
+        }
+
+        foreach ($this->client->stream($this->responses, $remainingTimeout) as $response => $chunk) {
+            try {
+                if ($chunk->isTimeout() && !$errorOnTimeout) {
+                    continue;
+                }
+
+                if (!$chunk->isFirst() && !$chunk->isLast()) {
+                    continue;
+                }
+
+                [$onHeaders, $onBody] = $this->responses[$response];
+                $onHeaders = $onHeaders ?? $this->onHeaders;
+                $onBody = $onBody ?? $this->onBody;
+
+                if (null !== $onHeaders && $chunk->isFirst()) {
+                    $onHeaders($response);
+                }
+
+                if (null !== $onBody && $chunk->isLast()) {
+                    $onBody($response);
+                }
+
+                if (null === $onBody || $chunk->isLast()) {
```

Comment:
why checking `null === onBody` ? if it's the last chunk, it should be removed from the responses stack.
IMHO, there is a leak here